### ISSUE TITLE
Enhancement: api optimization

### ISSUE THREAD

**Issue by Ilya-Green** [jowilf/starlette-admin#236]
Is there any method to exclude unnecessary fields from "/api/{identity}" repsone? (entries from other pages, unnecessary fields)
It responds a lot of not displayed information and even sensitive. 
it has a very long response on not very big database with relations.
I've tried to modify "find_all" method, but i didn't find way to exclude unnecessary field from  already generated sqlalchemy statment.
I don't understand how "get_list_query()" works either, it returns very complicated class.
My attempts to form statement by myself lead to errors in other places. It seems that custom select class creates something that need for other program components.

also i didnt realy understant what this code making. It append relations field and increases the sql query several times but relation fields already included in statement (it slows down request by more than 2 times):
```
async def find_all(
...
        for field in self.fields:
            if isinstance(field, RelationField):
                stmt = stmt.options(joinedload(getattr(self.model, field.name)))
```

**Comment by jowilf** [comment]
Please provide more information by filling the [feature request template](https://github.com/jowilf/starlette-admin/issues/new?assignees=&labels=enhancement&template=feature_request.md&title=Enhancement%3A+) so I can properly address the issue. Thank you

**Comment by Ilya-Green** [comment]
**Is your feature request related to a problem? Please describe.**
When you open the page /admin/{model}/list, an API fetch request is made to (hypothetically) /admin/api/{model}?skip=0&limit=20&order_by=id%20asc. This request is where the whole problem lies. It returns a lot of unnecessary data.

1. It retrieves all records instead of only those displayed on the page.
2. It displays fields that are excluded from the list view.
3. It fetches all related model instances and all related model instances of those related models and so on.

This leads to the following issue: if all models in the database are connected, opening any page retrieves all records from the database of all models at once.

Moreover, even sensitive data that is hidden from viewing and editing is displayed in API response.

**Describe the solution you'd like**
I tried to edit sqla\view.py: def "find all." This function constructs the SQL query and executes it. However, I couldn't achieve the desired outcome. The main issue is that I couldn't find a way to exclude unnecessary fields from the statement. Also, rewriting the logic for forming the query was unsuccessful because a custom statement class is used as a base, and I don't understand how it works. When creating an instance, it immediately includes those unnecessary fields.

**Comment by jowilf** [comment]
Methods for Fetching Data:

1. `get_list_query(self) -> Select`:
   - This method returns a SQLAlchemy `Select` expression, serving as the base statement for the `find_all` method. By default, it uses `select(Model)` to fetch all columns from the model.
   - You can override this method to control which columns are selected or to add additional WHERE clauses without directly modifying the `find_all` method.

2. `find_all(self, request: Request, skip: int = 0, limit: int = 100, where: Union[Dict[str, Any], str, None] = None, order_by: Optional[List[str]] = None) -> Sequence[Any]`:
   - This method is responsible for fetching a list of records from the database based on the provided parameters. It utilizes the `get_list_query` method as the base statement and applies filtering(search builder or full-text search ), pagination (`skip` and `limit`), and ordering (`order_by`) based on the provided arguments.
   - Eager Loading: It eagerly loads related models specified in the `fields` attribute.
```python
for field in self.fields:
    if isinstance(field, RelationField):
        stmt = stmt.options(joinedload(getattr(self.model, field.name)))
```

3. `get_count_query(self) -> Select`:
   - This method returns a SQLAlchemy `Select` expression, serving as the base statement for the `count` method. By default, it's set to `select(func.count(primary_key))`.
   - If you override the `get_list_query` method and add additional WHERE clauses, ensure to add the same conditions here to display the correct item count in the list view.

4. `count(self, request: Request, where: Union[Dict[str, Any], str, None] = None) -> int`:
   - This method counts the total number of records. It uses the `get_count_query` method as the base statement

Serialization for Frontend:

After fetching the necessary data using the `find_all` method, the retrieved records are serialized to JSON before being sent to the frontend.  Only `_edit_url` and `_detail_url` are added as a meta information. The serialization process only includes the fields specified in the `fields` property.

https://github.com/jowilf/starlette-admin/blob/8d242352c25b5d2e60e88d528c551f65668e0141/starlette_admin/views.py#L475


**Note:** As you can see the code does the opposite of what you described. If you are indeed experiencing the problems you mentioned, you might consider filing a bug report

**Comment by Ilya-Green** [comment]
limit parametr didnt affect on anything
![image](https://github.com/jowilf/starlette-admin/assets/105509043/7d162f7c-9e78-4e6c-9f86-da2432e2b8e2)

**Comment by Ilya-Green** [comment]
This is difference between enabled and disabled this part of code 
```
        for field in self.fields:
            if isinstance(field, RelationField):
                stmt = stmt.options(joinedload(getattr(self.model, field.name)))
```
![image](https://github.com/jowilf/starlette-admin/assets/105509043/a255f1b5-aa1d-47eb-8489-165fdc6fb79c)
![image](https://github.com/jowilf/starlette-admin/assets/105509043/7c7b73ba-008f-47c5-abb5-864a513f9a74)
Response of server is the same and on functionality of crm it didn't affects (i have a lot of relations) but time of execution is can be 9th time slower (lite on screen)

**Comment by jowilf** [comment]
> Can you please show example of `get_list_query(self) -> Select` of excluding/showing specific fields? I can't find any way to do this. These are my attempts, they didn't work:
> 
> 1. `self.get_list_query().select(Client.id, Client.first_name)` returns all fields in api return anyway
>    I didn't see any methods where i can define fields or exclude in self.get_list_query()
> 2. If i'm trying to sqlalchemy base select class, this leads to error
>    statement:
>    `stmt = select(User.id, User.first_name)`
>    traceback:
>    `AttributeError: 'int' object has no attribute 'id'`

You can use the `fields` property of the ModelView class to customize which fields are included in the admin view. https://jowilf.github.io/starlette-admin/tutorial/configurations/modelview/#fields. 

The `total` count is computed with the `count()` methods which generates a `SELECT COUNT(id) FROM ....` query. You can put the limit at 1 to see how many items is returned by the api.

It seems that there is a missing check for eager loading of RelationField -> `and not field.exclude_fields_from_list`

```python
for field in self.fields:
    if isinstance(field, RelationField) and not field.exclude_fields_from_list:
        stmt = stmt.options(joinedload(getattr(self.model, field.name)))
```
However if you exclude the field from the admin view (by removing it from the list of `fields` -> [See](https://jowilf.github.io/starlette-admin/tutorial/configurations/modelview/#fields.)) it will not be included 


it would be helpful if you could share a minimal code

**Comment by Ilya-Green** [comment]
> You can use the fields property of the ModelView class to customize which fields are included in the admin view.

But if i want to exclude only from list view. It can be record that have thousands records with relation to him and every of them will be executed and all of them information. I dont want to show this information on list view and dont want to execute it all every list view. But want to have option to edit.
It can be some sensitive data that i didnt want to close for editing for someone who has edit access.


> The total count is computed with the count() methods which generates a SELECT COUNT(id) FROM .... query. You can put the limit at 1 to see how many items is returned by the api.

Total count number is not the issue. The problem is that there is no need to fetch all possible thousands records just to draw 20 of them on frontend. And then fetch all of them again to draw the second page to display the next 20 records.


```
    async def find_all(
        self,
        request: Request,
        skip: int = 0,
        limit: int = 100,
        where: Union[Dict[str, Any], str, None] = None,
        order_by: Optional[List[str]] = None,
    ) -> Sequence[Any]:
        session: Union[Session, AsyncSession] = request.state.session
        stmt = self.get_list_query().offset(skip)
        if limit > 0:
            stmt = stmt.limit(limit)
        if where is not None:
            if isinstance(where, dict):
                where = build_query(where, self.model)
            else:
                where = await self.build_full_text_search_query(
                    request, where, self.model
                )
            stmt = stmt.where(where)  # type: ignore
        stmt = stmt.order_by(*build_order_clauses(order_by or [], self.model))

>         for field in self.fields and not field.exclude_fields_from_list:
>             if isinstance(field, RelationField) and not field.exclude_fields_from_list:
>                 stmt = stmt.options(joinedload(getattr(self.model, field.name)))

        if isinstance(session, AsyncSession):
            return (await session.execute(stmt)).scalars().unique().all()
        return (
            (await anyio.to_thread.run_sync(session.execute, stmt))
            .scalars()
            .unique()
            .all()
        )
```

`AttributeError: 'HasOne' object has no attribute 'exclude_fields_from_list'`
At least this error occurred on sqlmodel version.

> it would be helpful if you could share a minimal code

Ok, I'll create an example and send it

**Comment by jowilf** [comment]
> AttributeError: 'HasOne' object has no attribute 'exclude_fields_from_list'

Sorry, The correct attribute to use is `exclude_from_edit`, not `exclude_fields_from_list`. This draft pull request #251 fixed it. Could you please confirm whether this fix resolves your problem?

> The problem is that there is no need to fetch all possible thousands records just to draw 20 of them on frontend. 

As mentioned earlier `starlette-admin` apply a [limit](https://docs.sqlalchemy.org/en/20/orm/queryguide/query.html#sqlalchemy.orm.Query.limit) and [offset](https://docs.sqlalchemy.org/en/20/orm/queryguide/query.html#sqlalchemy.orm.Query.offset) according to sqlalchemy documentation. If all the data is being fetched for each page in your case, you can submit a bug report with more detail. To investigate further and understand the queries being executed, you can add `echo=True` to SQLAlchemy's engine configuration. This will log the SQL requests, allowing you to see if the limit and offset clauses are properly applied in the generated queries.

**Comment by jowilf** [comment]
@Ilya-Green could you confirm whether your issue is solved or not?

**Comment by jowilf** [comment]
I believe the issue has been resolved in version 0.11.1.  Feel free to reopen if the problem persists
